kanidmd/src/lib/entry.rs

```diff
@@ -1462,28 +1462,72 @@ impl Entry<EntryReduced, EntryCommitted> {
         audit: &mut AuditScope,
         qs: &QueryServerReadTransaction,
         basedn: &str,
+        // Did the client request all attributes?
+        all_attrs: bool,
+        // Did the ldap client request any sperific attribute names? If so,
+        // we need to remap everything to match.
+        l_attrs: &[String],
     ) -> Result<LdapSearchResultEntry, OperationError> {
         let rdn = qs.uuid_to_rdn(audit, self.get_uuid())?;
 
         let dn = format!("{},{}", rdn, basedn);
 
-        let attributes: Result<Vec<_>, _> = self
+        // Everything in our attrs set is "what was requested". So we can transform that now
+        // so they are all in "ldap forms" which makes our next stage a bit easier.
+
+        // Stage 1 - transform our results to a map of kani attr -> ldap value.
+        let attr_map: Result<Map<&str, Vec<String>>, _> = self
             .attrs
             .iter()
             .map(|(k, vs)| {
                 let pvs: Result<Vec<String>, _> = vs
                     .iter()
                     .map(|v| qs.resolve_value_ldap(audit, v, basedn))
                     .collect();
-                let pvs = pvs?;
-                let ks = ldap_attr_entry_map(k.as_str());
-                Ok(LdapPartialAttribute {
-                    atype: ks,
-                    vals: pvs,
-                })
+                pvs.map(|pvs| (k.as_str(), pvs))
+            })
+            .collect();
+
+        let attr_map = attr_map?;
+        // Stage 2 - transform and get all our attr - names out that we need to return.
+        //                  ldap a, kani a
+        let attr_names: Vec<(&str, &str)> = if all_attrs {
+            // Join the set of attr keys, and our requested attrs.
+            self.attrs
+                .keys()
+                .map(|k| (k.as_str(), k.as_str()))
+                .chain(
+                    l_attrs
```

Comment:
This means that every search you always get keys AND ssh_publickeys back. Should we not generate these virtual attrs in a default search?  And only use them when requested?

Reply:
In the update I think I have a reasonable compromise. If you req '*' you just get kanidm attrs, if you req '+' you get all kanidm attrs + all mapped attributes so you can see what mappings exist. And then you can request "what you like".